[quant][pyper] make embedding_bag quantization static

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44048 [quant][pyper] Support quantization of ops in fork-wait subgraph
* **#44008 [quant][pyper] make embedding_bag quantization static**
* #43989 [quant][pyper] Support aten::embedding_bag quantization in graph mode

Summary:
embedding_bag requires only quantization of weights (no dynamic quantization of inputs)
So the type of quantization is essentially static (without calibration)
This will enable pyper to do fc and embedding_bag quantization using the same API call

Test Plan:
python test/test_quantization.py test_embedding_bag

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23467019](https://our.internmc.facebook.com/intern/diff/D23467019)

supriyar committed:
[quant][pyper] make embedding_bag quantization static
Summary:
embedding_bag requires only quantization of weights (no dynamic quantization of inputs)
So the type of quantization is essentially static (without calibration)
This will enable pyper to do fc and embedding_bag quantization using the same API call

Test Plan:
python test/test_quantization.py test_embedding_bag

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][pyper] make embedding_bag quantization static"


Summary:
embedding_bag requires only quantization of weights (no dynamic quantization of inputs)
So the type of quantization is essentially static (without calibration)
This will enable pyper to do fc and embedding_bag quantization using the same API call

Test Plan:
python test/test_quantization.py test_embedding_bag

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23467019](https://our.internmc.facebook.com/intern/diff/D23467019)

[ghstack-poisoned]

supriyar committed:
Update on "[quant][pyper] make embedding_bag quantization static"


Summary:
embedding_bag requires only quantization of weights (no dynamic quantization of inputs)
So the type of quantization is essentially static (without calibration)
This will enable pyper to do fc and embedding_bag quantization using the same API call

Test Plan:
python test/test_quantization.py test_embedding_bag

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23467019](https://our.internmc.facebook.com/intern/diff/D23467019)

[ghstack-poisoned]

raghuramank100 commented on test/quantization/test_quantize_jit.py:
```diff
@@ -2928,14 +2929,14 @@ def forward(self, indices1, offsets1, indices2, offsets2):
+            int8_qconfig = QConfig(activation=PlaceholderObserver.with_args(dtype=torch.float,
+                                                                            custom_op_name="embedding_bag_byte"),
+                                   weight=PlaceholderObserver.with_args(custom_op_name="embedding_bag_byte"))
+            m = prepare_jit(m, {'embedding1' : int4_qconfig, 'embedding2' : int8_qconfig})
```
What about eager mode? Currently we expose embedding bag as a module in torch/nn/quantized/dynamic. Should we change that too?. Strictly speaking this case straddles the boundary of static vs dynamic:  Output activations are in fp32 (like dynamic). Inputs are addresses and require no quantization.  Weights alone are quantized.

supriyar replied:
I'm not sure about eager mode. Currently for static quant we expect the users to provide calibration fn when they call `quantize`. Weight only quantization shouldn't require that step. So in that sense it fits better into dynamic.

supriyar committed:
Update on "[quant][pyper] make embedding_bag quantization static"


Summary:
embedding_bag requires only quantization of weights (no dynamic quantization of inputs)
So the type of quantization is essentially static (without calibration)
This will enable pyper to do fc and embedding_bag quantization using the same API call

Test Plan:
python test/test_quantization.py test_embedding_bag

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23467019](https://our.internmc.facebook.com/intern/diff/D23467019)

[ghstack-poisoned]
